### PR TITLE
Fixing component description to consider local config and cluster

### LINKED ISSUE #2778: 'odo describe' does not include 'not pushed' URLs into output
/kind bug

## What versions of software are you using?

**Operating System:**
Windows 10
**Output of `odo version`:**
odo v1.1.1 (02fbd3d08)

## How did you run odo exactly?
1. Created odo git nodejs component
2. Added two urls
3. Pushed to the cluster
4. `odo describe` prints out both URLs
5. `odo url create url3` 

## Actual behavior
`odo describe` still prints the same info as at step (4) no url3 int the list of URLs.

## Expected behavior
`odo describe` reports all URLs pushed and not pushed.

## Any logs, error output, etc?
N/A

### PR DESCRIPTION
**Pull request by mohammedzee1000**
- created new struct
- structurally similar to component but url is no longer list of strings
  but URLList

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

/kind bug

**Which issue(s) this PR fixes**:

Fixes #2778 

**Testing**
- Build odo from this 
- Create a component
- do a `odo describe` with and without json output
- add URLs, storages, with various combinations of pushed and unpushed
- try describe again

**NOTE**: Not yet implemented for devfile

### CODE REVIEW

File: pkg/odo/cli/component/describe.go [hunk 1]
```diff
@@ -56,31 +56,20 @@ func (do *DescribeOptions) Validate() (err error) {
...
+	}
 
-	if state == component.StateTypeNotPushed || state == component.StateTypeUnknown {
-		componentDesc, err = component.GetComponentFromConfig(do.LocalConfigInfo)
-		componentDesc.Status.State = state
-		if err != nil {
-			return err
-		}
+	if log.IsJSON() {
+		cfd.Spec.Ports = do.LocalConfigInfo.GetPorts()
```
Comment (mohammedzee1000) [inline pkg/odo/cli/component/describe.go:65]: Just noticed. This line is no longer required. Will remove it after the reviews

File: pkg/component/component_full_description.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	state := GetComponentState(client, componentName, applicationName)
+	componentDesc, err := GetComponentFromConfig(localConfigInfo)
+	if err != nil {
+		return cfd, err
+	}
+	err = cfd.copyFromComponentDesc(&componentDesc)
+	if err != nil {
+		return cfd, err
+	}
+	cfd.Status.State = state
```
Comment (mohammedzee1000) [inline pkg/component/component_full_description.go:111]: this too damn why do I notice this after asking for review :(

File: pkg/odo/cli/component/describe.go (hunk 1 above)
Comment (mohammedzee1000) [reply pkg/odo/cli/component/describe.go]: Done

**Comment by mohammedzee1000** [comment]
Failures unrelated
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2962/pull-ci-openshift-odo-master-v4.2-integration-e2e-benchmark/2315
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2962/pull-ci-openshift-odo-master-v4.4-integration-e2e-benchmark/451
/retest

**Comment by GeekArthur** [comment]
@mohammedzee1000 Seems like most of the code is regarding s2i component, it would great to let @cdrage or other Red Hat folks to review, they have better s2i component knowledge than me.

**Comment by mohammedzee1000** [comment]
@GeekArthur your comments are still going to be helpful as it will affect implementation for devfiles too. There is a related issue

**Comment by GeekArthur** [comment]
@mohammedzee1000 From scalability and compatibility perspective, I think the new struct would be good for devfile for now as currently the devfile only needs basic component metadata and the new struct covers.

For future I think we still need to do some changes if it's necessary because of two things:
1. Devfile schema https://github.com/redhat-developer/devfile is more complicated and we would decide if we need add them/part of them to the struct
2. Devfile 2.0 design is not finalized yet, we may need to do some changes to accommodate that as well

But I don't think things above should be finish or affect this PR, so it should be good now.

**Comment by mohammedzee1000** [comment]
@GeekArthur thanks for your comments and feedback

@cdrage review plz

**Comment by girishramnani** [comment]
We can get a sample describe json outputs for experimental and non experimental modes

**Comment by mohammedzee1000** [comment]
With experimental mode off
```
❯ odo describe -o json
{
    "kind": "Component",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "name": "nodejs-nodejs-ex-2-rwbv",
        "namespace": "mz-temp",
        "creationTimestamp": null
    },
    "spec": {
        "app": "app",
        "type": "nodejs",
        "source": "file://./",
        "sourceType": "local",
        "urls": {
            "kind": "List",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {},
            "items": [
                {
                    "kind": "url",
                    "apiVersion": "odo.openshift.io/v1alpha1",
                    "metadata": {
                        "name": "url1",
                        "creationTimestamp": null
                    },
                    "spec": {
                        "host": "url1-app-mz-temp.apps.testocp4x.psiodo.net",
                        "protocol": "http",
                        "port": 8080,
                        "secure": false
                    },
                    "status": {
                        "state": "Pushed"
                    }
                },
                {
                    "kind": "url",
                    "apiVersion": "odo.openshift.io/v1alpha1",
                    "metadata": {
                        "name": "url2",
                        "creationTimestamp": null
                    },
                    "spec": {
                        "port": 8080,
                        "secure": false
                    },
                    "status": {
                        "state": "Not Pushed"
                    }
                }
            ]
        },
        "storages": {
            "kind": "List",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {},
            "items": null
        },
        "ports": [
            "8080/TCP"
        ]
    },
    "status": {
        "state": "Pushed"
    }
}


```

Have not yet tried with experimental mode

**Comment by mohammedzee1000** [comment]
With experimental mode on existing normal component:
```
❯ odo describe
Component Name: nodejs-nodejs-ex-2-rwbv
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858

❯ odo describe -o json
{
    "kind": "Component",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "name": "nodejs-nodejs-ex-2-rwbv",
        "namespace": "mz-temp",
        "creationTimestamp": null
    },
    "spec": {
        "app": "app",
        "type": "nodejs",
        "source": "file://./",
        "sourceType": "local",
        "urls": {
            "kind": "List",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {},
            "items": [
                {
                    "kind": "url",
                    "apiVersion": "odo.openshift.io/v1alpha1",
                    "metadata": {
                        "name": "url1",
                        "creationTimestamp": null
                    },
                    "spec": {
                        "host": "url1-app-mz-temp.apps.testocp4x.psiodo.net",
                        "protocol": "http",
                        "port": 8080,
                        "secure": false
                    },
                    "status": {
                        "state": "Pushed"
                    }
                },
                {
                    "kind": "url",
                    "apiVersion": "odo.openshift.io/v1alpha1",
                    "metadata": {
                        "name": "url2",
                        "creationTimestamp": null
                    },
                    "spec": {
                        "port": 8080,
                        "secure": false
                    },
                    "status": {
                        "state": "Not Pushed"
                    }
                }
            ]
        },
        "storages": {
            "kind": "List",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {},
            "items": null
        },
        "env": [
            {
                "name": "DEBUG_PORT",
                "value": "5858"
            }
        ],
        "ports": [
            "8080/TCP"
        ]
    },
    "status": {
        "state": "Pushed"
    }
}


```

**Comment by mohammedzee1000** [comment]
experimental mode, new component

```
❯ odo create nodejs
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile compatibility [153902ns]
 ✓  Validating devfile component [362750ns]

Please use `odo push` command to create the component with source deployed
❯ ls
devfile.yaml  package.json  server.js

❯ cat devfile.yaml
apiVersion: 1.0.0
metadata:
  generateName: nodejs-
projects:
  -
    name: nodejs-web-app
    source:
      type: git
      location: "https://github.com/che-samples/web-nodejs-sample.git"
components:
  - type: dockerimage
    alias: runtime
    image: quay.io/eclipse/che-nodejs10-ubi:nightly
    memoryLimit: 1024Mi
    mountSources: true
    endpoints:
      - name: "3000/tcp"
        port: 3000
commands:
  - name: devBuild
    actions:
      - type: exec
        component: runtime
        command: npm install
        workdir: ${CHE_PROJECTS_ROOT}/nodejs-web-app/app
  - name: devRun
    actions:
      - type: exec
        component: runtime
        command: nodemon app.js
        workdir: ${CHE_PROJECTS_ROOT}/nodejs-web-app/app

❯ odo describe
 ✗  The current directory does not represent an odo component. Use 'odo create' to create component here or switch to directory with a component
❯ odo describe -o json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "The current directory does not represent an odo component. Use 'odo create' to create component here or switch to directory with a component"
}

```

**Comment by mohammedzee1000** [comment]
Again like geekauther said, experimental mode will not be fixed in this PR. There will be more work that needs to be done. It does solve the issue that started this and lay the groundwork to fix bigger problem

**Comment by mohammedzee1000** [comment]
@girishramnani ^

**Comment by girishramnani** [comment]
Can we please update the PR description to show how to reproduce these json outputs?

**Comment by mohammedzee1000** [comment]
@girishramnani that ok?
@cdrage review plz

File: pkg/component/component_full_description.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/occlient"
+	"github.com/openshift/odo/pkg/odo/util/experimental"
+	"github.com/openshift/odo/pkg/storage"
+	urlpkg "github.com/openshift/odo/pkg/url"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//ComponentFullDescriptionSpec repersents complete desciption of the component
```
Comment (cdrage) [inline pkg/component/component_full_description.go:18]: Space after comments `//`

File: pkg/component/component_full_description.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+			return err
+		}
+		storages = storage.ConvertListLocalToMachine(storageLocal)
+	}
+	cfd.Spec.Storage = storages
+	return nil
+}
+
+func (cfd *ComponentFullDescription) fillEmptyFields(componentDesc Component, componentName string, applicationName string, projectName string) {
+	//fix missing names in case it in not in description
```
Comment (cdrage) [inline pkg/component/component_full_description.go:76]: space after `//`

File: pkg/component/component_full_description.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+		cfd.APIVersion = "odo.openshift.io/v1alpha1"
+	}
+
+	if len(cfd.Spec.App) <= 0 {
+		cfd.Spec.App = applicationName
+	}
+	cfd.Spec.Ports = componentDesc.Spec.Ports
+}
+
+//NewComponentFullDescriptionFromClientAndLocalConfig gets the complete description of the component from both localconfig and cluster
```
Comment (cdrage) [inline pkg/component/component_full_description.go:99]: space after `//`

File: pkg/component/component_full_description.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	urls, err := urlpkg.List(client, localConfigInfo, componentName, applicationName)
+	if err != nil {
+		return err
+	}
+
+	cfd.Spec.URL = urls
+	return nil
+}
+
+func (cfd *ComponentFullDescription) loadStoragesFromClientAndLocalConfig(client *occlient.Client, localConfigInfo *config.LocalConfigInfo, componentName string, applicationName string, componentDesc *Component) error {
```
Comment (cdrage) [inline pkg/component/component_full_description.go:56]: I feel as though this function should have more comments describing how stuff works similar to your `NewComponentFullDescriptionFromClientAndLocalConfig` function.

File: pkg/component/component_full_description.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	err = cfd.loadStoragesFromClientAndLocalConfig(client, localConfigInfo, componentName, applicationName, &componentDesc)
+	if err != nil {
+		return cfd, err
+	}
+	return cfd, nil
+}
+
+//Print prints the complete information of component onto stdout (Note: longterm this function should not need to access any parameters, but just print the information in struct)
```
Comment (cdrage) [inline pkg/component/component_full_description.go:136]: space after `//`

File: pkg/component/component_full_description.go (hunk 3 above)
Comment (mohammedzee1000) [reply pkg/component/component_full_description.go:18]: fixed, will push shortly

File: pkg/component/component_full_description.go (hunk 4 above)
Comment (mohammedzee1000) [reply pkg/component/component_full_description.go:76]: fixed

File: pkg/component/component_full_description.go (hunk 5 above)
Comment (mohammedzee1000) [reply pkg/component/component_full_description.go:99]: fixed

File: pkg/component/component_full_description.go (hunk 7 above)
Comment (mohammedzee1000) [reply pkg/component/component_full_description.go:136]: fixed

File: pkg/component/component_full_description.go (hunk 6 above)
Comment (mohammedzee1000) [reply pkg/component/component_full_description.go:56]: added

**Pull request merged**